From 8dcbbd4576dd02bff955516288dae1ab6262babb Mon Sep 17 00:00:00 2001
From: Abhay Singh Gill <abhaygill017@gmail.com>
Date: Sat, 5 Sep 2026 15:17:27 +0530
Subject: [PATCH 22/30] base: Add key generation for keybox spoofing [1/2]

The code is used from :-
https://github.com/5ec1cff/TrickyStore

This allows us to enjoy spoofing keyboxes the right way even with broken TEE.
Allow users to provide keybox XML as well.

Change-Id: I0da036c907f8cd1f6d2e3480cf3d5f688ac17dc0
Signed-off-by: Abhay Singh Gill <abhaygill017@gmail.com>
---
 core/java/android/provider/Settings.java      |   7 +
 .../internal/util/KeyProviderManager.java     | 104 -----
 .../internal/util/KeyboxImitationHooks.java   | 211 ---------
 .../internal/util/PropImitationHooks.java     |   1 +
 .../util/{ => extra}/IKeyboxProvider.java     |   2 +-
 .../util/extra/KeyProviderManager.java        | 198 ++++++++
 .../util/extra/KeyboxChainGenerator.java      | 437 ++++++++++++++++++
 .../util/extra/KeyboxImitationHooks.java      | 172 +++++++
 .../internal/util/extra/KeyboxUtils.java      | 154 ++++++
 keystore/java/android/security/KeyStore2.java |  10 +-
 .../security/KeyStoreSecurityLevel.java       |  20 +
 .../validators/SecureSettingsValidators.java  |   1 +
 12 files changed, 998 insertions(+), 319 deletions(-)
 delete mode 100644 core/java/com/android/internal/util/KeyProviderManager.java
 delete mode 100644 core/java/com/android/internal/util/KeyboxImitationHooks.java
 rename core/java/com/android/internal/util/{ => extra}/IKeyboxProvider.java (96%)
 create mode 100644 core/java/com/android/internal/util/extra/KeyProviderManager.java
 create mode 100644 core/java/com/android/internal/util/extra/KeyboxChainGenerator.java
 create mode 100644 core/java/com/android/internal/util/extra/KeyboxImitationHooks.java
 create mode 100644 core/java/com/android/internal/util/extra/KeyboxUtils.java

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 36e2193cec11..10179e9f5f23 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -13334,6 +13334,13 @@ public final class Settings {
         public static final String HBM_SETTING_KEY =
                 "com.android.server.display.HBM_SETTING_KEY";
 
+        /**
+         * User selectable keybox data.
+         * @hide
+         */
+        @Readable
+        public static final String KEYBOX_DATA = "keybox_data";
+
         /**
          * Keys we no longer back up under the current schema, but want to continue to
          * process when restoring historical backup datasets.
diff --git a/core/java/com/android/internal/util/KeyProviderManager.java b/core/java/com/android/internal/util/KeyProviderManager.java
deleted file mode 100644
index 2e0b7f8e17f2..000000000000
--- a/core/java/com/android/internal/util/KeyProviderManager.java
+++ /dev/null
@@ -1,104 +0,0 @@
-/*
- * SPDX-FileCopyrightText: 2024 Paranoid Android
- * SPDX-License-Identifier: Apache-2.0
- */
-package com.android.internal.util;
-
-import android.app.ActivityThread;
-import android.content.Context;
-import android.util.Log;
-
-import com.android.internal.R;
-
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.Map;
-
-/**
- * Manager class for handling keybox providers.
- * @hide
- */
-public final class KeyProviderManager {
-    private static final String TAG = "KeyProviderManager";
-
-    private static final IKeyboxProvider PROVIDER = new DefaultKeyboxProvider();
-
-    private KeyProviderManager() {
-    }
-
-    public static IKeyboxProvider getProvider() {
-        return PROVIDER;
-    }
-
-    public static boolean isKeyboxAvailable() {
-        return PROVIDER.hasKeybox();
-    }
-
-    private static class DefaultKeyboxProvider implements IKeyboxProvider {
-        private final Map<String, String> keyboxData = new HashMap<>();
-
-        private DefaultKeyboxProvider() {
-            Context context = getApplicationContext();
-            if (context == null) {
-                Log.e(TAG, "Failed to get application context");
-                return;
-            }
-
-            String[] keybox = context.getResources().getStringArray(R.array.config_certifiedKeybox);
-
-            Arrays.stream(keybox)
-                    .map(entry -> entry.split(":", 2))
-                    .filter(parts -> parts.length == 2)
-                    .forEach(parts -> keyboxData.put(parts[0], parts[1]));
-
-            if (!hasKeybox()) {
-                Log.w(TAG, "Incomplete keybox data loaded");
-            }
-        }
-
-        private static Context getApplicationContext() {
-            try {
-                return ActivityThread.currentApplication().getApplicationContext();
-            } catch (Exception e) {
-                Log.e(TAG, "Error getting application context", e);
-                return null;
-            }
-        }
-
-        @Override
-        public boolean hasKeybox() {
-            return Arrays.asList("EC.PRIV", "EC.CERT_1", "EC.CERT_2", "EC.CERT_3",
-                    "RSA.PRIV", "RSA.CERT_1", "RSA.CERT_2", "RSA.CERT_3")
-                    .stream()
-                    .allMatch(keyboxData::containsKey);
-        }
-
-        @Override
-        public String getEcPrivateKey() {
-            return keyboxData.get("EC.PRIV");
-        }
-
-        @Override
-        public String getRsaPrivateKey() {
-            return keyboxData.get("RSA.PRIV");
-        }
-
-        @Override
-        public String[] getEcCertificateChain() {
-            return getCertificateChain("EC");
-        }
-
-        @Override
-        public String[] getRsaCertificateChain() {
-            return getCertificateChain("RSA");
-        }
-
-        private String[] getCertificateChain(String prefix) {
-            return new String[]{
-                    keyboxData.get(prefix + ".CERT_1"),
-                    keyboxData.get(prefix + ".CERT_2"),
-                    keyboxData.get(prefix + ".CERT_3")
-            };
-        }
-    }
-}
diff --git a/core/java/com/android/internal/util/KeyboxImitationHooks.java b/core/java/com/android/internal/util/KeyboxImitationHooks.java
deleted file mode 100644
index 9b2e95dfa2c3..000000000000
--- a/core/java/com/android/internal/util/KeyboxImitationHooks.java
+++ /dev/null
@@ -1,211 +0,0 @@
-/*
- * SPDX-FileCopyrightText: 2024 Paranoid Android
- * SPDX-License-Identifier: Apache-2.0
- */
-package com.android.internal.util;
-
-import android.os.SystemProperties;
-import android.security.KeyChain;
-import android.security.keystore.KeyProperties;
-import android.system.keystore2.KeyEntryResponse;
-import android.util.Log;
-
-import com.android.internal.org.bouncycastle.asn1.ASN1Boolean;
-import com.android.internal.org.bouncycastle.asn1.ASN1Encodable;
-import com.android.internal.org.bouncycastle.asn1.ASN1EncodableVector;
-import com.android.internal.org.bouncycastle.asn1.ASN1Enumerated;
-import com.android.internal.org.bouncycastle.asn1.ASN1ObjectIdentifier;
-import com.android.internal.org.bouncycastle.asn1.ASN1OctetString;
-import com.android.internal.org.bouncycastle.asn1.ASN1Sequence;
-import com.android.internal.org.bouncycastle.asn1.ASN1TaggedObject;
-import com.android.internal.org.bouncycastle.asn1.DEROctetString;
-import com.android.internal.org.bouncycastle.asn1.DERSequence;
-import com.android.internal.org.bouncycastle.asn1.DERTaggedObject;
-import com.android.internal.org.bouncycastle.asn1.x509.Extension;
-import com.android.internal.org.bouncycastle.cert.X509CertificateHolder;
-import com.android.internal.org.bouncycastle.cert.X509v3CertificateBuilder;
-import com.android.internal.org.bouncycastle.operator.ContentSigner;
-import com.android.internal.org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
-
-import java.io.ByteArrayOutputStream;
-import java.security.KeyFactory;
-import java.security.PrivateKey;
-import java.security.cert.X509Certificate;
-import java.security.spec.PKCS8EncodedKeySpec;
-import java.util.Base64;
-import java.util.concurrent.ThreadLocalRandom;
-
-/**
- * @hide
- */
-public class KeyboxImitationHooks {
-
-    private static final String TAG = "KeyboxImitationHooks";
-    private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
-
-    private static final Boolean sDisableKeyAttestationBlock = SystemProperties.getBoolean(
-            "persist.sys.pihooks.disable.gms_key_attestation_block", false);
-
-    private static final ASN1ObjectIdentifier KEY_ATTESTATION_OID = new ASN1ObjectIdentifier(
-            "1.3.6.1.4.1.11129.2.1.17");
-
-    private static PrivateKey parsePrivateKey(String encodedKey, String algorithm)
-            throws Exception {
-        byte[] keyBytes = Base64.getDecoder().decode(encodedKey);
-        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(keyBytes);
-        return KeyFactory.getInstance(algorithm).generatePrivate(keySpec);
-    }
-
-    private static byte[] parseCertificate(String encodedCert) {
-        return Base64.getDecoder().decode(encodedCert);
-    }
-
-    private static byte[] getCertificateChain(String algorithm) throws Exception {
-        IKeyboxProvider provider = KeyProviderManager.getProvider();
-        String[] certChain = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
-                ? provider.getEcCertificateChain()
-                : provider.getRsaCertificateChain();
-
-        ByteArrayOutputStream certificateStream = new ByteArrayOutputStream();
-        for (String cert : certChain) {
-            certificateStream.write(parseCertificate(cert));
-        }
-        return certificateStream.toByteArray();
-    }
-
-    private static PrivateKey getPrivateKey(String algorithm) throws Exception {
-        IKeyboxProvider provider = KeyProviderManager.getProvider();
-        String privateKeyEncoded = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
-                ? provider.getEcPrivateKey()
-                : provider.getRsaPrivateKey();
-
-        return parsePrivateKey(privateKeyEncoded, algorithm);
-    }
-
-    private static X509CertificateHolder getCertificateHolder(String algorithm) throws Exception {
-        IKeyboxProvider provider = KeyProviderManager.getProvider();
-        String certChain = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
-                ? provider.getEcCertificateChain()[0]
-                : provider.getRsaCertificateChain()[0];
-
-        return new X509CertificateHolder(parseCertificate(certChain));
-    }
-
-    private static byte[] modifyLeafCertificate(X509Certificate leafCertificate,
-            String keyAlgorithm) throws Exception {
-        X509CertificateHolder certificateHolder = new X509CertificateHolder(
-                leafCertificate.getEncoded());
-        Extension keyAttestationExtension = certificateHolder.getExtension(KEY_ATTESTATION_OID);
-        ASN1Sequence keyAttestationSequence = ASN1Sequence.getInstance(
-                keyAttestationExtension.getExtnValue().getOctets());
-        ASN1Encodable[] keyAttestationEncodables = keyAttestationSequence.toArray();
-        ASN1Sequence teeEnforcedSequence = (ASN1Sequence) keyAttestationEncodables[7];
-        ASN1EncodableVector teeEnforcedVector = new ASN1EncodableVector();
-
-        ASN1Sequence rootOfTrustSequence = null;
-        for (ASN1Encodable teeEnforcedEncodable : teeEnforcedSequence) {
-            ASN1TaggedObject taggedObject = (ASN1TaggedObject) teeEnforcedEncodable;
-            if (taggedObject.getTagNo() == 704) {
-                rootOfTrustSequence = (ASN1Sequence) taggedObject.getObject();
-                continue;
-            }
-            teeEnforcedVector.add(teeEnforcedEncodable);
-        }
-
-        if (rootOfTrustSequence == null) throw new Exception("Root of trust not found");
-
-        PrivateKey privateKey = getPrivateKey(keyAlgorithm);
-        X509CertificateHolder providerCertHolder = getCertificateHolder(keyAlgorithm);
-
-        X509v3CertificateBuilder certificateBuilder = new X509v3CertificateBuilder(
-                providerCertHolder.getSubject(),
-                certificateHolder.getSerialNumber(),
-                certificateHolder.getNotBefore(),
-                certificateHolder.getNotAfter(),
-                certificateHolder.getSubject(),
-                certificateHolder.getSubjectPublicKeyInfo()
-        );
-
-        ContentSigner contentSigner = new JcaContentSignerBuilder(
-                leafCertificate.getSigAlgName()).build(privateKey);
-
-        byte[] verifiedBootKey = new byte[32];
-        ThreadLocalRandom.current().nextBytes(verifiedBootKey);
-
-        DEROctetString verifiedBootHash = (DEROctetString) rootOfTrustSequence.getObjectAt(3);
-        if (verifiedBootHash == null) {
-            byte[] randomHash = new byte[32];
-            ThreadLocalRandom.current().nextBytes(randomHash);
-            verifiedBootHash = new DEROctetString(randomHash);
-        }
-
-        ASN1Encodable[] rootOfTrustEncodables = {
-                new DEROctetString(verifiedBootKey),
-                ASN1Boolean.TRUE,
-                new ASN1Enumerated(0),
-                verifiedBootHash
-        };
-
-        ASN1Sequence newRootOfTrustSequence = new DERSequence(rootOfTrustEncodables);
-        ASN1TaggedObject rootOfTrustTaggedObject = new DERTaggedObject(704, newRootOfTrustSequence);
-        teeEnforcedVector.add(rootOfTrustTaggedObject);
-
-        ASN1Sequence newTeeEnforcedSequence = new DERSequence(teeEnforcedVector);
-        keyAttestationEncodables[7] = newTeeEnforcedSequence;
-        ASN1Sequence newKeyAttestationSequence = new DERSequence(keyAttestationEncodables);
-        ASN1OctetString newKeyAttestationOctetString = new DEROctetString(
-                newKeyAttestationSequence);
-        Extension newKeyAttestationExtension = new Extension(KEY_ATTESTATION_OID, false,
-                newKeyAttestationOctetString);
-
-        certificateBuilder.addExtension(newKeyAttestationExtension);
-
-        for (ASN1ObjectIdentifier extensionOID :
-                certificateHolder.getExtensions().getExtensionOIDs()) {
-            if (KEY_ATTESTATION_OID.getId().equals(extensionOID.getId())) continue;
-            certificateBuilder.addExtension(certificateHolder.getExtension(extensionOID));
-        }
-
-        return certificateBuilder.build(contentSigner).getEncoded();
-    }
-
-    public static KeyEntryResponse onGetKeyEntry(KeyEntryResponse response) {
-        if (sDisableKeyAttestationBlock) {
-            dlog("Key attestation spoofing is disabled by user");
-            return response;
-        }
-
-        // If no keybox is found, don't continue spoofing
-        if (!KeyProviderManager.isKeyboxAvailable()) {
-            dlog("Key attestation spoofing is disabled because no keybox is defined to spoof");
-            return response;
-        }
-
-        if (response == null || response.metadata == null) return response;
-
-        try {
-            if (response.metadata.certificate == null) {
-                Log.e(TAG, "Certificate is null, skipping modification");
-                return response;
-            }
-
-            X509Certificate certificate = KeyChain.toCertificate(response.metadata.certificate);
-            if (certificate.getExtensionValue(KEY_ATTESTATION_OID.getId()) == null) {
-                Log.e(TAG, "Key attestation OID not found, skipping modification");
-                return response;
-            }
-
-            String keyAlgorithm = certificate.getPublicKey().getAlgorithm();
-            response.metadata.certificate = modifyLeafCertificate(certificate, keyAlgorithm);
-            response.metadata.certificateChain = getCertificateChain(keyAlgorithm);
-        } catch (Exception e) {
-            Log.e(TAG, "Error in onGetKeyEntry", e);
-        }
-
-        return response;
-    }
-
-    private static void dlog(String msg) {
-        if (DEBUG) Log.d(TAG, msg);
-    }
-}
diff --git a/core/java/com/android/internal/util/PropImitationHooks.java b/core/java/com/android/internal/util/PropImitationHooks.java
index 8699af4c39b2..b025cfc58ff7 100644
--- a/core/java/com/android/internal/util/PropImitationHooks.java
+++ b/core/java/com/android/internal/util/PropImitationHooks.java
@@ -32,6 +32,7 @@ import android.text.TextUtils;
 import android.util.Log;
 
 import com.android.internal.R;
+import com.android.internal.util.extra.KeyProviderManager;
 
 import java.lang.reflect.Field;
 import java.util.Arrays;
diff --git a/core/java/com/android/internal/util/IKeyboxProvider.java b/core/java/com/android/internal/util/extra/IKeyboxProvider.java
similarity index 96%
rename from core/java/com/android/internal/util/IKeyboxProvider.java
rename to core/java/com/android/internal/util/extra/IKeyboxProvider.java
index 3463fcce572d..ee285f1275e3 100644
--- a/core/java/com/android/internal/util/IKeyboxProvider.java
+++ b/core/java/com/android/internal/util/extra/IKeyboxProvider.java
@@ -2,7 +2,7 @@
  * SPDX-FileCopyrightText: 2024 Paranoid Android
  * SPDX-License-Identifier: Apache-2.0
  */
-package com.android.internal.util;
+package com.android.internal.util.extra;
 
 /**
  * Interface for keybox providers.
diff --git a/core/java/com/android/internal/util/extra/KeyProviderManager.java b/core/java/com/android/internal/util/extra/KeyProviderManager.java
new file mode 100644
index 000000000000..69f39850a46a
--- /dev/null
+++ b/core/java/com/android/internal/util/extra/KeyProviderManager.java
@@ -0,0 +1,198 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Paranoid Android
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.android.internal.util.extra;
+
+import android.app.ActivityThread;
+import android.content.Context;
+import android.provider.Settings;
+import android.util.Log;
+import android.util.Xml;
+
+import com.android.internal.R;
+import org.xmlpull.v1.XmlPullParser;
+
+import java.io.StringReader;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Manager class for handling keybox providers.
+ * @hide
+ */
+public final class KeyProviderManager {
+    private static final String TAG = "KeyProviderManager";
+
+    private KeyProviderManager() {}
+
+    public static IKeyboxProvider getProvider() {
+        return new DefaultKeyboxProvider();
+    }
+
+    public static boolean isKeyboxAvailable() {
+        return getProvider().hasKeybox();
+    }
+
+    private static class DefaultKeyboxProvider implements IKeyboxProvider {
+        private final Map<String, String> keyboxData = new HashMap<>();
+
+        private DefaultKeyboxProvider() {
+            Context context = getApplicationContext();
+            if (context == null) {
+                Log.e(TAG, "Failed to get application context");
+                return;
+            }
+
+            if (!loadFromXmlSetting(context)) {
+                loadFromConfigArray(context);
+            }
+        }
+
+        private boolean loadFromXmlSetting(Context ctx) {
+            try {
+                String xml = Settings.Secure.getString(ctx.getContentResolver(), Settings.Secure.KEYBOX_DATA);
+                if (xml == null || xml.trim().isEmpty()) return false;
+
+                XmlPullParser p = Xml.newPullParser();
+                p.setInput(new StringReader(xml));
+
+                String currentAlg = null;
+                int certCount = 0;
+                boolean numberOfKeyboxesChecked = false;
+
+                for (int ev = p.next(); ev != XmlPullParser.END_DOCUMENT; ev = p.next()) {
+                    if (ev == XmlPullParser.START_TAG) {
+                        String tag = p.getName();
+                        switch (tag) {
+                            case "NumberOfKeyboxes":
+                                p.next();
+                                numberOfKeyboxesChecked = true;
+                                try {
+                                    int count = Integer.parseInt(p.getText().trim());
+                                    if (count != 1) {
+                                        Log.w(TAG, "Invalid NumberOfKeyboxes: " + count);
+                                        return false;
+                                    }
+                                } catch (NumberFormatException e) {
+                                    Log.w(TAG, "Failed to parse NumberOfKeyboxes", e);
+                                    return false;
+                                }
+                                break;
+
+                            case "Key":
+                                currentAlg = p.getAttributeValue(null, "algorithm");
+                                if ("ecdsa".equalsIgnoreCase(currentAlg)) currentAlg = "EC";
+                                else if ("rsa".equalsIgnoreCase(currentAlg)) currentAlg = "RSA";
+                                else currentAlg = null;
+                                certCount = 0;
+                                break;
+
+                            case "PrivateKey": {
+                                String format = p.getAttributeValue(null, "format");
+                                if (!"pem".equalsIgnoreCase(format)) {
+                                    Log.w(TAG, "Unsupported PrivateKey format: " + format);
+                                    return false;
+                                }
+                                p.next();
+                                if (currentAlg != null) {
+                                    keyboxData.put(currentAlg + ".PRIV", p.getText().trim());
+                                }
+                                break;
+                            }
+
+                            case "Certificate": {
+                                String format = p.getAttributeValue(null, "format");
+                                if (!"pem".equalsIgnoreCase(format)) {
+                                    Log.w(TAG, "Unsupported Certificate format: " + format);
+                                    return false;
+                                }
+                                if (currentAlg != null && certCount < 3) {
+                                    p.next();
+                                    certCount++;
+                                    keyboxData.put(currentAlg + ".CERT_" + certCount, p.getText().trim());
+                                }
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                if (!numberOfKeyboxesChecked) {
+                    Log.w(TAG, "Missing <NumberOfKeyboxes> in keybox XML");
+                    return false;
+                }
+
+                if (!hasKeybox()) {
+                    Log.w(TAG, "Failed to load keybox from XML setting");
+                    return false;
+                }
+
+                Log.i(TAG, "Loaded keybox from XML setting");
+                return true;
+            } catch (Exception e) {
+                Log.e(TAG, "XML keybox load failed", e);
+                return false;
+            }
+        }
+
+        private void loadFromConfigArray(Context ctx) {
+            for (String entry : ctx.getResources().getStringArray(R.array.config_certifiedKeybox)) {
+                String[] parts = entry.split(":", 2);
+                if (parts.length == 2) {
+                    keyboxData.put(parts[0], parts[1]);
+                }
+            }
+
+            if (!hasKeybox()) {
+                Log.w(TAG, "Incomplete keybox provided by overlays");
+            }
+        }
+
+        private static Context getApplicationContext() {
+            try {
+                return ActivityThread.currentApplication().getApplicationContext();
+            } catch (Exception e) {
+                Log.e(TAG, "Error getting application context", e);
+                return null;
+            }
+        }
+
+        @Override
+        public boolean hasKeybox() {
+            return Arrays.asList("EC.PRIV", "EC.CERT_1", "EC.CERT_2", "EC.CERT_3",
+                    "RSA.PRIV", "RSA.CERT_1", "RSA.CERT_2", "RSA.CERT_3")
+                    .stream()
+                    .allMatch(keyboxData::containsKey);
+        }
+
+        @Override
+        public String getEcPrivateKey() {
+            return keyboxData.get("EC.PRIV");
+        }
+
+        @Override
+        public String getRsaPrivateKey() {
+            return keyboxData.get("RSA.PRIV");
+        }
+
+        @Override
+        public String[] getEcCertificateChain() {
+            return getCertificateChain("EC");
+        }
+
+        @Override
+        public String[] getRsaCertificateChain() {
+            return getCertificateChain("RSA");
+        }
+
+        private String[] getCertificateChain(String prefix) {
+            return new String[]{
+                    keyboxData.get(prefix + ".CERT_1"),
+                    keyboxData.get(prefix + ".CERT_2"),
+                    keyboxData.get(prefix + ".CERT_3")
+            };
+        }
+    }
+}
diff --git a/core/java/com/android/internal/util/extra/KeyboxChainGenerator.java b/core/java/com/android/internal/util/extra/KeyboxChainGenerator.java
new file mode 100644
index 000000000000..02940f1a7ca5
--- /dev/null
+++ b/core/java/com/android/internal/util/extra/KeyboxChainGenerator.java
@@ -0,0 +1,437 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Neoteric OS
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.android.internal.util.extra;
+
+import android.app.ActivityThread;
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.content.pm.Signature;
+import android.hardware.security.keymint.Algorithm;
+import android.hardware.security.keymint.EcCurve;
+import android.hardware.security.keymint.KeyParameter;
+import android.hardware.security.keymint.Tag;
+import android.os.Binder;
+import android.os.Build;
+import android.security.keystore.KeyProperties;
+import android.system.keystore2.KeyDescriptor;
+import android.util.Log;
+
+import androidx.annotation.Nullable;
+
+import com.android.internal.org.bouncycastle.asn1.ASN1Boolean;
+import com.android.internal.org.bouncycastle.asn1.ASN1Encodable;
+import com.android.internal.org.bouncycastle.asn1.ASN1Enumerated;
+import com.android.internal.org.bouncycastle.asn1.ASN1Integer;
+import com.android.internal.org.bouncycastle.asn1.ASN1ObjectIdentifier;
+import com.android.internal.org.bouncycastle.asn1.ASN1OctetString;
+import com.android.internal.org.bouncycastle.asn1.ASN1Sequence;
+import com.android.internal.org.bouncycastle.asn1.DERNull;
+import com.android.internal.org.bouncycastle.asn1.DEROctetString;
+import com.android.internal.org.bouncycastle.asn1.DERSequence;
+import com.android.internal.org.bouncycastle.asn1.DERSet;
+import com.android.internal.org.bouncycastle.asn1.DERTaggedObject;
+import com.android.internal.org.bouncycastle.asn1.x500.X500Name;
+import com.android.internal.org.bouncycastle.asn1.x509.Extension;
+import com.android.internal.org.bouncycastle.asn1.x509.KeyUsage;
+import com.android.internal.org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import com.android.internal.org.bouncycastle.asn1.x509.Time;
+import com.android.internal.org.bouncycastle.cert.X509CertificateHolder;
+import com.android.internal.org.bouncycastle.cert.X509v3CertificateBuilder;
+import com.android.internal.org.bouncycastle.jce.provider.BouncyCastleProvider;
+import com.android.internal.org.bouncycastle.operator.ContentSigner;
+import com.android.internal.org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.security.Security;
+import java.security.cert.Certificate;
+import java.security.spec.ECGenParameterSpec;
+import java.security.spec.RSAKeyGenParameterSpec;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import javax.security.auth.x500.X500Principal;
+
+/**
+ * @hide
+ */
+public final class KeyboxChainGenerator {
+
+    private static final String TAG = "KeyboxChainGenerator";
+    private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
+
+    private static final int ATTESTATION_APPLICATION_ID_PACKAGE_INFOS_INDEX = 0;
+    private static final int ATTESTATION_APPLICATION_ID_SIGNATURE_DIGESTS_INDEX = 1;
+    private static final int ATTESTATION_PACKAGE_INFO_PACKAGE_NAME_INDEX = 0;
+    private static final int ATTESTATION_PACKAGE_INFO_VERSION_INDEX = 1;
+
+    public static List<Certificate> generateCertChain(int uid, KeyDescriptor descriptor, KeyGenParameters params) {
+        dlog("Requested KeyPair with alias: " + descriptor.alias);
+        int size = params.keySize;
+        KeyPair kp;
+        try {
+            if (Objects.equals(params.algorithm, Algorithm.EC)) {
+                dlog("Generating EC keypair of size " + size);
+                kp = buildECKeyPair(params);
+            } else if (Objects.equals(params.algorithm, Algorithm.RSA)) {
+                dlog("Generating RSA keypair of size " + size);
+                kp = buildRSAKeyPair(params);
+            } else {
+                dlog("Unsupported algorithm");
+                return null;
+            }
+
+            X509v3CertificateBuilder certBuilder = new X509v3CertificateBuilder(
+                    KeyboxUtils.getCertificateHolder(
+                            Objects.equals(params.algorithm, Algorithm.EC)
+                                    ? KeyProperties.KEY_ALGORITHM_EC
+                                    : KeyProperties.KEY_ALGORITHM_RSA
+                    ).getSubject(),
+                    params.certificateSerial,
+                    new Time(params.certificateNotBefore),
+                    new Time(params.certificateNotAfter),
+                    params.certificateSubject,
+                    SubjectPublicKeyInfo.getInstance(
+                            ASN1Sequence.getInstance(kp.getPublic().getEncoded())
+                    )
+            );
+
+            KeyUsage keyUsage = new KeyUsage(KeyUsage.keyCertSign);
+            certBuilder.addExtension(Extension.keyUsage, true, keyUsage);
+            certBuilder.addExtension(createExtension(params, uid));
+
+            ContentSigner contentSigner;
+            if (Objects.equals(params.algorithm, Algorithm.EC)) {
+                contentSigner = new JcaContentSignerBuilder("SHA256withECDSA").build(KeyboxUtils.getPrivateKey(KeyProperties.KEY_ALGORITHM_EC));
+            } else {
+                contentSigner = new JcaContentSignerBuilder("SHA256withRSA").build(KeyboxUtils.getPrivateKey(KeyProperties.KEY_ALGORITHM_RSA));
+            }
+            X509CertificateHolder certHolder = certBuilder.build(contentSigner);
+            Certificate leaf = KeyboxUtils.getCertificateFromHolder(certHolder);
+            List<Certificate> chain = KeyboxUtils.getCertificateChain(leaf.getPublicKey().getAlgorithm());
+            chain.add(0, leaf);
+            dlog("Successfully generated X500 Cert for alias: " + descriptor.alias);
+            return chain;
+        } catch (Throwable t) {
+            Log.e(TAG, Log.getStackTraceString(t));
+        }
+        return null;
+    }
+
+    private static ASN1Encodable[] fromIntList(List<Integer> list) {
+        ASN1Encodable[] result = new ASN1Encodable[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            result[i] = new ASN1Integer(list.get(i));
+        }
+        return result;
+    }
+
+    private static Extension createExtension(KeyGenParameters params, int uid) {
+        try {
+            SecureRandom random = new SecureRandom();
+
+            byte[] bytes1 = new byte[32];
+            byte[] bytes2 = new byte[32];
+
+            random.nextBytes(bytes1);
+            random.nextBytes(bytes2);
+
+            ASN1Encodable[] rootOfTrustEncodables = {new DEROctetString(bytes1), ASN1Boolean.TRUE,
+                    new ASN1Enumerated(0), new DEROctetString(bytes2)};
+
+            ASN1Sequence rootOfTrustSeq = new DERSequence(rootOfTrustEncodables);
+
+            var Apurpose = new DERSet(fromIntList(params.purpose));
+            var Aalgorithm = new ASN1Integer(params.algorithm);
+            var AkeySize = new ASN1Integer(params.keySize);
+            var Adigest = new DERSet(fromIntList(params.digest));
+            var AecCurve = new ASN1Integer(params.ecCurve);
+            var AnoAuthRequired = DERNull.INSTANCE;
+
+            // To be loaded
+            var AosVersion = new ASN1Integer(getOsVersion());
+            var AosPatchLevel = new ASN1Integer(getPatchLevel());
+
+            var AapplicationID = createApplicationId(uid);
+            var AbootPatchlevel = new ASN1Integer(getPatchLevelLong());
+            var AvendorPatchLevel = new ASN1Integer(getPatchLevelLong());
+
+            var AcreationDateTime = new ASN1Integer(System.currentTimeMillis());
+            var Aorigin = new ASN1Integer(0);
+
+            var purpose = new DERTaggedObject(true, 1, Apurpose);
+            var algorithm = new DERTaggedObject(true, 2, Aalgorithm);
+            var keySize = new DERTaggedObject(true, 3, AkeySize);
+            var digest = new DERTaggedObject(true, 5, Adigest);
+            var ecCurve = new DERTaggedObject(true, 10, AecCurve);
+            var noAuthRequired = new DERTaggedObject(true, 503, AnoAuthRequired);
+            var creationDateTime = new DERTaggedObject(true, 701, AcreationDateTime);
+            var origin = new DERTaggedObject(true, 702, Aorigin);
+            var rootOfTrust = new DERTaggedObject(true, 704, rootOfTrustSeq);
+            var osVersion = new DERTaggedObject(true, 705, AosVersion);
+            var osPatchLevel = new DERTaggedObject(true, 706, AosPatchLevel);
+            var applicationID = new DERTaggedObject(true, 709, AapplicationID);
+            var vendorPatchLevel = new DERTaggedObject(true, 718, AvendorPatchLevel);
+            var bootPatchLevel = new DERTaggedObject(true, 719, AbootPatchlevel);
+
+            ASN1Encodable[] teeEnforcedEncodables;
+
+            // Support device properties attestation
+            if (params.brand != null) {
+                var Abrand = new DEROctetString(params.brand);
+                var Adevice = new DEROctetString(params.device);
+                var Aproduct = new DEROctetString(params.product);
+                var Amanufacturer = new DEROctetString(params.manufacturer);
+                var Amodel = new DEROctetString(params.model);
+                var brand = new DERTaggedObject(true, 710, Abrand);
+                var device = new DERTaggedObject(true, 711, Adevice);
+                var product = new DERTaggedObject(true, 712, Aproduct);
+                var manufacturer = new DERTaggedObject(true, 716, Amanufacturer);
+                var model = new DERTaggedObject(true, 717, Amodel);
+
+                teeEnforcedEncodables = new ASN1Encodable[]{purpose, algorithm, keySize, digest, ecCurve,
+                        noAuthRequired, origin, rootOfTrust, osVersion, osPatchLevel, vendorPatchLevel,
+                        bootPatchLevel, brand, device, product, manufacturer, model};
+            } else {
+                teeEnforcedEncodables = new ASN1Encodable[]{purpose, algorithm, keySize, digest, ecCurve,
+                        noAuthRequired, origin, rootOfTrust, osVersion, osPatchLevel, vendorPatchLevel,
+                        bootPatchLevel};
+            }
+
+            ASN1Encodable[] softwareEnforced = {applicationID, creationDateTime};
+
+            ASN1OctetString keyDescriptionOctetStr = getAsn1OctetString(teeEnforcedEncodables, softwareEnforced, params);
+
+            return new Extension(new ASN1ObjectIdentifier("1.3.6.1.4.1.11129.2.1.17"), false, keyDescriptionOctetStr);
+        } catch (Throwable t) {
+            Log.e(TAG, Log.getStackTraceString(t));
+        }
+        return null;
+    }
+
+    private static int getOsVersion() {
+        String release = Build.VERSION.RELEASE;
+        int major = 0, minor = 0, patch = 0;
+
+        String[] parts = release.split("\\.");
+        if (parts.length > 0) major = Integer.parseInt(parts[0]);
+        if (parts.length > 1) minor = Integer.parseInt(parts[1]);
+        if (parts.length > 2) patch = Integer.parseInt(parts[2]);
+
+        return major * 10000 + minor * 100 + patch;
+    }
+
+    private static int getPatchLevel() {
+        return convertPatchLevel(Build.VERSION.SECURITY_PATCH, false);
+    }
+
+    private static int getPatchLevelLong() {
+        return convertPatchLevel(Build.VERSION.SECURITY_PATCH, true);
+    }
+
+    private static int convertPatchLevel(String patchLevel, boolean longFormat) {
+        try {
+            String[] parts = patchLevel.split("-");
+            int year = Integer.parseInt(parts[0]);
+            int month = Integer.parseInt(parts[1]);
+            if (longFormat) {
+                int day = Integer.parseInt(parts[2]);
+                return year * 10000 + month * 100 + day;
+            } else {
+                return year * 100 + month;
+            }
+        } catch (Exception e) {
+            Log.e(TAG, "Invalid patch level: " + patchLevel, e);
+            return 202404;
+        }
+    }
+
+    private static ASN1OctetString getAsn1OctetString(ASN1Encodable[] teeEnforcedEncodables, ASN1Encodable[] softwareEnforcedEncodables, KeyGenParameters params) throws IOException {
+        ASN1Integer attestationVersion = new ASN1Integer(100);
+        ASN1Enumerated attestationSecurityLevel = new ASN1Enumerated(1);
+        ASN1Integer keymasterVersion = new ASN1Integer(100);
+        ASN1Enumerated keymasterSecurityLevel = new ASN1Enumerated(1);
+        ASN1OctetString attestationChallenge = new DEROctetString(params.attestationChallenge);
+        ASN1OctetString uniqueId = new DEROctetString("".getBytes());
+        ASN1Encodable softwareEnforced = new DERSequence(softwareEnforcedEncodables);
+        ASN1Sequence teeEnforced = new DERSequence(teeEnforcedEncodables);
+
+        ASN1Encodable[] keyDescriptionEncodables = {attestationVersion, attestationSecurityLevel, keymasterVersion,
+                keymasterSecurityLevel, attestationChallenge, uniqueId, softwareEnforced, teeEnforced};
+
+        ASN1Sequence keyDescriptionHackSeq = new DERSequence(keyDescriptionEncodables);
+
+        return new DEROctetString(keyDescriptionHackSeq);
+    }
+
+    private static DEROctetString createApplicationId(int uid) throws Throwable {
+        Context context = ActivityThread.currentApplication();
+        if (context == null) {
+            throw new IllegalStateException("createApplicationId: context not available from ActivityThread!");
+        }
+
+        PackageManager pm = context.getPackageManager();
+        if (pm == null) {
+            throw new IllegalStateException("createApplicationId: PackageManager not found!");
+        }
+
+        String[] packages = pm.getPackagesForUid(uid);
+        if (packages == null || packages.length == 0) {
+            throw new IllegalStateException("No packages found for UID: " + uid);
+        }
+
+        int size = packages.length;
+        ASN1Encodable[] packageInfoAA = new ASN1Encodable[size];
+        Set<Digest> signatures = new HashSet<>();
+        MessageDigest dg = MessageDigest.getInstance("SHA-256");
+
+        for (int i = 0; i < size; i++) {
+            String name = packages[i];
+            PackageInfo info = pm.getPackageInfo(name, PackageManager.GET_SIGNATURES);
+            ASN1Encodable[] arr = new ASN1Encodable[2];
+            arr[ATTESTATION_PACKAGE_INFO_PACKAGE_NAME_INDEX] =
+                    new DEROctetString(name.getBytes(StandardCharsets.UTF_8));
+            arr[ATTESTATION_PACKAGE_INFO_VERSION_INDEX] =
+                    new ASN1Integer(info.getLongVersionCode());
+            packageInfoAA[i] = new DERSequence(arr);
+
+            for (Signature s : info.signatures) {
+                signatures.add(new Digest(dg.digest(s.toByteArray())));
+            }
+        }
+
+        ASN1Encodable[] signaturesAA = new ASN1Encodable[signatures.size()];
+        int i = 0;
+        for (Digest d : signatures) {
+            signaturesAA[i++] = new DEROctetString(d.digest);
+        }
+
+        ASN1Encodable[] applicationIdAA = new ASN1Encodable[2];
+        applicationIdAA[ATTESTATION_APPLICATION_ID_PACKAGE_INFOS_INDEX] =
+                new DERSet(packageInfoAA);
+        applicationIdAA[ATTESTATION_APPLICATION_ID_SIGNATURE_DIGESTS_INDEX] =
+                new DERSet(signaturesAA);
+
+        return new DEROctetString(new DERSequence(applicationIdAA).getEncoded());
+    }
+
+    record Digest(byte[] digest) {
+        @Override
+        public boolean equals(@Nullable Object o) {
+            if (o instanceof Digest d)
+                return Arrays.equals(digest, d.digest);
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return Arrays.hashCode(digest);
+        }
+    }
+
+    private static KeyPair buildECKeyPair(KeyGenParameters params) throws Exception {
+        Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME);
+        Security.addProvider(new BouncyCastleProvider());
+        ECGenParameterSpec spec = new ECGenParameterSpec(params.ecCurveName);
+        KeyPairGenerator kpg = KeyPairGenerator.getInstance("EC", BouncyCastleProvider.PROVIDER_NAME);
+        kpg.initialize(spec);
+        return kpg.generateKeyPair();
+    }
+
+    private static KeyPair buildRSAKeyPair(KeyGenParameters params) throws Exception {
+        Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME);
+        Security.addProvider(new BouncyCastleProvider());
+        RSAKeyGenParameterSpec spec = new RSAKeyGenParameterSpec(
+                params.keySize, params.rsaPublicExponent);
+        KeyPairGenerator kpg = KeyPairGenerator.getInstance("RSA", BouncyCastleProvider.PROVIDER_NAME);
+        kpg.initialize(spec);
+        return kpg.generateKeyPair();
+    }
+
+    private static void dlog(String msg) {
+        if (DEBUG) Log.d(TAG, msg);
+    }
+
+    public static class KeyGenParameters {
+        public int keySize;
+        public int algorithm;
+        public BigInteger certificateSerial;
+        public Date certificateNotBefore;
+        public Date certificateNotAfter;
+        public X500Name certificateSubject;
+
+        public BigInteger rsaPublicExponent;
+        public int ecCurve;
+        public String ecCurveName;
+
+        public List<Integer> purpose = new ArrayList<>();
+        public List<Integer> digest = new ArrayList<>();
+
+        public byte[] attestationChallenge;
+        public byte[] brand;
+        public byte[] device;
+        public byte[] product;
+        public byte[] manufacturer;
+        public byte[] model;
+
+        public int securityLevel;
+
+        public KeyGenParameters(KeyParameter[] params) {
+            for (KeyParameter kp : params) {
+                switch (kp.tag) {
+                    case Tag.KEY_SIZE -> keySize = kp.value.getInteger();
+                    case Tag.ALGORITHM -> algorithm = kp.value.getAlgorithm();
+                    case Tag.CERTIFICATE_SERIAL -> certificateSerial = new BigInteger(kp.value.getBlob());
+                    case Tag.CERTIFICATE_NOT_BEFORE -> certificateNotBefore = new Date(kp.value.getDateTime());
+                    case Tag.CERTIFICATE_NOT_AFTER -> certificateNotAfter = new Date(kp.value.getDateTime());
+                    case Tag.CERTIFICATE_SUBJECT -> certificateSubject =
+                            new X500Name(new X500Principal(kp.value.getBlob()).getName());
+                    case Tag.RSA_PUBLIC_EXPONENT -> rsaPublicExponent = BigInteger.valueOf(kp.value.getLongInteger());
+                    case Tag.EC_CURVE -> {
+                        ecCurve = kp.value.getEcCurve();
+                        ecCurveName = getEcCurveName(ecCurve);
+                    }
+                    case Tag.PURPOSE -> {
+                        purpose.add(kp.value.getKeyPurpose());
+                    }
+                    case Tag.DIGEST -> {
+                        digest.add(kp.value.getDigest());
+                    }
+                    case Tag.ATTESTATION_CHALLENGE -> attestationChallenge = kp.value.getBlob();
+                    case Tag.ATTESTATION_ID_BRAND -> brand = kp.value.getBlob();
+                    case Tag.ATTESTATION_ID_DEVICE -> device = kp.value.getBlob();
+                    case Tag.ATTESTATION_ID_PRODUCT -> product = kp.value.getBlob();
+                    case Tag.ATTESTATION_ID_MANUFACTURER -> manufacturer = kp.value.getBlob();
+                    case Tag.ATTESTATION_ID_MODEL -> model = kp.value.getBlob();
+                    case Tag.HARDWARE_TYPE -> securityLevel = kp.value.getSecurityLevel();
+                }
+            }
+        }
+
+        private static String getEcCurveName(int curve) {
+            String res;
+            switch (curve) {
+                case EcCurve.CURVE_25519 -> res = "CURVE_25519";
+                case EcCurve.P_224 -> res = "secp224r1";
+                case EcCurve.P_256 -> res = "secp256r1";
+                case EcCurve.P_384 -> res = "secp384r1";
+                case EcCurve.P_521 -> res = "secp521r1";
+                default -> throw new IllegalArgumentException("unknown curve");
+            }
+            return res;
+        }
+    }
+}
diff --git a/core/java/com/android/internal/util/extra/KeyboxImitationHooks.java b/core/java/com/android/internal/util/extra/KeyboxImitationHooks.java
new file mode 100644
index 000000000000..518d179055af
--- /dev/null
+++ b/core/java/com/android/internal/util/extra/KeyboxImitationHooks.java
@@ -0,0 +1,172 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Paranoid Android
+ * SPDX-FileCopyrightText: 2025 Neoteric OS
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.android.internal.util.extra;
+
+import android.hardware.security.keymint.Algorithm;
+import android.hardware.security.keymint.KeyParameter;
+import android.hardware.security.keymint.KeyParameterValue;
+import android.hardware.security.keymint.Tag;
+import android.os.Binder;
+import android.system.keystore2.Authorization;
+import android.system.keystore2.IKeystoreSecurityLevel;
+import android.system.keystore2.KeyDescriptor;
+import android.system.keystore2.KeyEntryResponse;
+import android.system.keystore2.KeyMetadata;
+import android.util.Log;
+
+import com.android.internal.util.extra.KeyboxChainGenerator.KeyGenParameters;
+
+import java.security.cert.Certificate;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * @hide
+ */
+public class KeyboxImitationHooks {
+
+    private static final String TAG = "KeyboxImitationHooks";
+    private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
+    private static boolean mSuccess = false;
+
+    public static KeyEntryResponse onGetKeyEntry(KeyDescriptor descriptor) {
+        if (!KeyProviderManager.isKeyboxAvailable()) {
+            return null;
+        }
+
+        if (!mSuccess) {
+            return null;
+        }
+
+        KeyEntryResponse spoofed = KeyboxUtils.retrieve(Binder.getCallingUid(), descriptor.alias);
+        if (spoofed != null) {
+            dlog("Key entry spoofed");
+            return spoofed;
+        }
+
+        return null;
+    }
+
+    public static KeyMetadata generateKey(IKeystoreSecurityLevel level, KeyDescriptor descriptor, Collection<KeyParameter> args) {
+        if (!KeyProviderManager.isKeyboxAvailable()) {
+            return null;
+        }
+
+        KeyGenParameters params = new KeyGenParameters(args.toArray(new KeyParameter[args.size()]));
+        if (params.algorithm != Algorithm.EC && params.algorithm != Algorithm.RSA) {
+            Log.w(TAG, "Unsupported algorithm: " + params.algorithm);
+            return null;
+        }
+
+        int uid = Binder.getCallingUid();
+        try {
+            List<Certificate> chain = KeyboxChainGenerator.generateCertChain(uid, descriptor, params);
+            if (chain == null || chain.isEmpty()) {
+                return null;
+            }
+            KeyEntryResponse response = buildResponse(level, chain, params, descriptor);
+            if (response == null) {
+                return null;
+            }
+            KeyboxUtils.append(uid, descriptor.alias, response);
+            mSuccess = true;
+            return response.metadata;
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to generate key", e);
+            return null;
+        }
+    }
+
+    private static KeyEntryResponse buildResponse(
+            IKeystoreSecurityLevel level,
+            List<Certificate> chain,
+            KeyGenParameters params,
+            KeyDescriptor descriptor
+    ) {
+        try {
+            KeyEntryResponse response = new KeyEntryResponse();
+            KeyMetadata metadata = new KeyMetadata();
+            metadata.keySecurityLevel = params.securityLevel;
+
+            KeyboxUtils.putCertificateChain(metadata, chain.toArray(new Certificate[chain.size()]));
+
+            KeyDescriptor d = new KeyDescriptor();
+            d.domain = descriptor.domain;
+            d.nspace = descriptor.nspace;
+            metadata.key = d;
+
+            List<Authorization> authorizations = new ArrayList<>();
+            Authorization a;
+
+            for (Integer i : params.purpose) {
+                a = new Authorization();
+                a.keyParameter = new KeyParameter();
+                a.keyParameter.tag = Tag.PURPOSE;
+                a.keyParameter.value = KeyParameterValue.keyPurpose(i);
+                a.securityLevel = params.securityLevel;
+                authorizations.add(a);
+            }
+
+            for (Integer i : params.digest) {
+                a = new Authorization();
+                a.keyParameter = new KeyParameter();
+                a.keyParameter.tag = Tag.DIGEST;
+                a.keyParameter.value = KeyParameterValue.digest(i);
+                a.securityLevel = params.securityLevel;
+                authorizations.add(a);
+            }
+
+            a = new Authorization();
+            a.keyParameter = new KeyParameter();
+            a.keyParameter.tag = Tag.ALGORITHM;
+            a.keyParameter.value = KeyParameterValue.algorithm(params.algorithm);
+            a.securityLevel = params.securityLevel;
+            authorizations.add(a);
+
+            a = new Authorization();
+            a.keyParameter = new KeyParameter();
+            a.keyParameter.tag = Tag.KEY_SIZE;
+            a.keyParameter.value = KeyParameterValue.integer(params.keySize);
+            a.securityLevel = params.securityLevel;
+            authorizations.add(a);
+
+            a = new Authorization();
+            a.keyParameter = new KeyParameter();
+            a.keyParameter.tag = Tag.EC_CURVE;
+            a.keyParameter.value = KeyParameterValue.ecCurve(params.ecCurve);
+            a.securityLevel = params.securityLevel;
+            authorizations.add(a);
+
+            a = new Authorization();
+            a.keyParameter = new KeyParameter();
+            a.keyParameter.tag = Tag.NO_AUTH_REQUIRED;
+            a.keyParameter.value = KeyParameterValue.boolValue(true); // TODO: copy
+            a.securityLevel = params.securityLevel;
+            authorizations.add(a);
+
+            // TODO: ORIGIN, OS_VERSION, OS_PATCHLEVEL, VENDOR_PATCHLEVEL, BOOT_PATCHLEVEL,
+            // CREATION_DATETIME, USER_ID
+
+            metadata.authorizations = authorizations.toArray(new Authorization[0]);
+            response.metadata = metadata;
+            response.iSecurityLevel = level;
+            return response;
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to build key entry response", e);
+            return null;
+        }
+    }
+
+    public static void setSuccessFlag(boolean flag) {
+        mSuccess = flag;
+    }
+
+    private static void dlog(String msg) {
+        if (DEBUG) Log.d(TAG, msg);
+    }
+}
diff --git a/core/java/com/android/internal/util/extra/KeyboxUtils.java b/core/java/com/android/internal/util/extra/KeyboxUtils.java
new file mode 100644
index 000000000000..924e22fb5b4a
--- /dev/null
+++ b/core/java/com/android/internal/util/extra/KeyboxUtils.java
@@ -0,0 +1,154 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Neoteric OS
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.android.internal.util.extra;
+
+import android.security.keystore.KeyProperties;
+import android.system.keystore2.KeyEntryResponse;
+import android.system.keystore2.KeyMetadata;
+
+import com.android.internal.org.bouncycastle.asn1.ASN1Sequence;
+import com.android.internal.org.bouncycastle.asn1.ASN1Primitive;
+import com.android.internal.org.bouncycastle.asn1.DERNull;
+import com.android.internal.org.bouncycastle.asn1.pkcs.PKCSObjectIdentifiers;
+import com.android.internal.org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import com.android.internal.org.bouncycastle.asn1.pkcs.RSAPrivateKey;
+import com.android.internal.org.bouncycastle.asn1.sec.ECPrivateKey;
+import com.android.internal.org.bouncycastle.asn1.x9.X9ObjectIdentifiers;
+import com.android.internal.org.bouncycastle.asn1.x509.AlgorithmIdentifier;
+import com.android.internal.org.bouncycastle.cert.X509CertificateHolder;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.security.KeyFactory;
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.RSAPrivateCrtKeySpec;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * @hide
+ */
+public class KeyboxUtils {
+
+    private static final ConcurrentHashMap<Key, KeyEntryResponse> response = new ConcurrentHashMap<>();
+    public static record Key(int uid, String alias) {}
+
+    public static byte[] decodePemOrBase64(String input) {
+        String base64 = input
+                .replaceAll("-----BEGIN [^-]+-----", "")
+                .replaceAll("-----END [^-]+-----", "")
+                .replaceAll("\\s+", "");
+        return Base64.getDecoder().decode(base64);
+    }
+
+    public static PrivateKey parsePrivateKey(String encodedKey, String algorithm) throws Exception {
+        byte[] keyBytes = decodePemOrBase64(encodedKey);
+        ASN1Primitive primitive = ASN1Primitive.fromByteArray(keyBytes);
+        if ("EC".equalsIgnoreCase(algorithm)) {
+            try {
+                // Try parsing as PKCS#8
+                PrivateKeyInfo info = PrivateKeyInfo.getInstance(primitive);
+                return KeyFactory.getInstance("EC").generatePrivate(new PKCS8EncodedKeySpec(info.getEncoded()));
+            } catch (Exception e) {
+                // Possibly SEC1 / PKCS#1 EC
+                ASN1Sequence seq = ASN1Sequence.getInstance(primitive);
+                ECPrivateKey ecPrivateKey = ECPrivateKey.getInstance(seq);
+                AlgorithmIdentifier algId = new AlgorithmIdentifier(X9ObjectIdentifiers.id_ecPublicKey, ecPrivateKey.getParameters());
+                PrivateKeyInfo privInfo = new PrivateKeyInfo(algId, ecPrivateKey);
+                PKCS8EncodedKeySpec pkcs8Spec = new PKCS8EncodedKeySpec(privInfo.getEncoded());
+                return KeyFactory.getInstance("EC").generatePrivate(pkcs8Spec);
+            }
+        } else if ("RSA".equalsIgnoreCase(algorithm)) {
+            try {
+                // Try parsing as PKCS#8
+                return KeyFactory.getInstance("RSA").generatePrivate(new PKCS8EncodedKeySpec(keyBytes));
+            } catch (Exception e) {
+                // Parse as PKCS#1
+                RSAPrivateKey rsaKey = RSAPrivateKey.getInstance(primitive);
+                AlgorithmIdentifier algId = new AlgorithmIdentifier(PKCSObjectIdentifiers.rsaEncryption, DERNull.INSTANCE);
+                PrivateKeyInfo privInfo = new PrivateKeyInfo(algId, rsaKey);
+                PKCS8EncodedKeySpec pkcs8Spec = new PKCS8EncodedKeySpec(privInfo.getEncoded());
+                return KeyFactory.getInstance("RSA").generatePrivate(pkcs8Spec);
+            }
+        } else {
+            throw new IllegalArgumentException("Unsupported algorithm: " + algorithm);
+        }
+    }
+
+    public static X509Certificate parseCertificate(String encodedCert) throws Exception {
+        byte[] certBytes = decodePemOrBase64(encodedCert);
+        return (X509Certificate) CertificateFactory
+                .getInstance("X.509")
+                .generateCertificate(new ByteArrayInputStream(certBytes));
+    }
+
+    public static List<Certificate> getCertificateChain(String algorithm) throws Exception {
+        IKeyboxProvider provider = KeyProviderManager.getProvider();
+        String[] certChainPem = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
+                ? provider.getEcCertificateChain()
+                : provider.getRsaCertificateChain();
+
+        CertificateFactory factory = CertificateFactory.getInstance("X.509");
+        List<Certificate> certs = new ArrayList<>();
+
+        for (String certPem : certChainPem) {
+            certs.add(parseCertificate(certPem));
+        }
+
+        return certs;
+    }
+
+    public static void putCertificateChain(KeyEntryResponse response, Certificate[] chain) throws Exception {
+        putCertificateChain(response.metadata, chain);
+    }
+
+    public static void putCertificateChain(KeyMetadata metadata, Certificate[] chain) throws Exception {
+        metadata.certificate = chain[0].getEncoded();
+        var output = new ByteArrayOutputStream();
+        for (int i = 1; i < chain.length; i++) {
+            output.write(chain[i].getEncoded());
+        }
+        metadata.certificateChain = output.toByteArray();
+    }
+
+    public static X509Certificate getCertificateFromHolder(X509CertificateHolder holder) throws Exception {
+        CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+        ByteArrayInputStream in = new ByteArrayInputStream(holder.getEncoded());
+        return (X509Certificate) certFactory.generateCertificate(in);
+    }
+
+    public static PrivateKey getPrivateKey(String algorithm) throws Exception {
+        IKeyboxProvider provider = KeyProviderManager.getProvider();
+        String privateKeyEncoded = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
+                ? provider.getEcPrivateKey()
+                : provider.getRsaPrivateKey();
+
+        return parsePrivateKey(privateKeyEncoded, algorithm);
+    }
+
+    public static X509CertificateHolder getCertificateHolder(String algorithm) throws Exception {
+        IKeyboxProvider provider = KeyProviderManager.getProvider();
+        String certPem = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
+                ? provider.getEcCertificateChain()[0]
+                : provider.getRsaCertificateChain()[0];
+
+        X509Certificate parsedCert = parseCertificate(certPem);
+        return new X509CertificateHolder(parsedCert.getEncoded());
+    }
+
+    public static void append(int uid, String a, KeyEntryResponse c) {
+        response.put(new Key(uid, a), c);
+    }
+
+    public static KeyEntryResponse retrieve(int uid, String a) {
+        return response.get(new Key(uid, a));
+    }
+}
diff --git a/keystore/java/android/security/KeyStore2.java b/keystore/java/android/security/KeyStore2.java
index 5ddbfcc23878..5466b5ff27f2 100644
--- a/keystore/java/android/security/KeyStore2.java
+++ b/keystore/java/android/security/KeyStore2.java
@@ -32,7 +32,7 @@ import android.system.keystore2.KeyEntryResponse;
 import android.system.keystore2.ResponseCode;
 import android.util.Log;
 
-import com.android.internal.util.KeyboxImitationHooks;
+import com.android.internal.util.extra.KeyboxImitationHooks;
 
 import java.util.Calendar;
 
@@ -285,8 +285,12 @@ public class KeyStore2 {
             throws KeyStoreException {
         StrictMode.noteDiskRead();
 
-        return KeyboxImitationHooks.onGetKeyEntry(
-                handleRemoteExceptionWithRetry((service) -> service.getKeyEntry(descriptor)));
+        KeyEntryResponse response = KeyboxImitationHooks.onGetKeyEntry(descriptor);
+        if (response != null) {
+            return response;
+        }
+
+        return handleRemoteExceptionWithRetry((service) -> service.getKeyEntry(descriptor));
     }
 
     /**
diff --git a/keystore/java/android/security/KeyStoreSecurityLevel.java b/keystore/java/android/security/KeyStoreSecurityLevel.java
index 6ab148a8b4ea..5f335a335d01 100644
--- a/keystore/java/android/security/KeyStoreSecurityLevel.java
+++ b/keystore/java/android/security/KeyStoreSecurityLevel.java
@@ -19,6 +19,7 @@ package android.security;
 import android.annotation.NonNull;
 import android.app.compat.CompatChanges;
 import android.hardware.security.keymint.KeyParameter;
+import android.hardware.security.keymint.Tag;
 import android.os.Binder;
 import android.os.RemoteException;
 import android.os.ServiceSpecificException;
@@ -33,6 +34,8 @@ import android.system.keystore2.KeyMetadata;
 import android.system.keystore2.ResponseCode;
 import android.util.Log;
 
+import com.android.internal.util.extra.KeyboxImitationHooks;
+
 import java.util.Calendar;
 import java.util.Collection;
 
@@ -146,6 +149,23 @@ public class KeyStoreSecurityLevel {
             throws KeyStoreException {
         StrictMode.noteDiskWrite();
 
+        byte[] attestationChallenge = null;
+        for (KeyParameter kp : args) {
+            if (kp.tag == Tag.ATTESTATION_CHALLENGE) {
+                attestationChallenge = kp.value.getBlob();
+                break;
+            }
+        }
+
+        KeyboxImitationHooks.setSuccessFlag(false);
+        if (attestationChallenge != null && attestationKey == null) {
+            KeyMetadata metadata = KeyboxImitationHooks.generateKey(mSecurityLevel,
+                    descriptor, args);
+            if (metadata != null) {
+                return metadata;
+            }
+        }
+
         return handleExceptions(() -> mSecurityLevel.generateKey(
                 descriptor, attestationKey, args.toArray(new KeyParameter[args.size()]),
                 flags, entropy));
diff --git a/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java b/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
index a7508c30f8c0..2e14ca580c58 100644
--- a/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
+++ b/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
@@ -521,5 +521,6 @@ public class SecureSettingsValidators {
                 new InclusiveIntegerRangeValidator(0, 1));
         VALIDATORS.put(Secure.IDENTITY_CHECK_NOTIFICATION_VIEW_DETAILS_CLICKED,
                 new InclusiveIntegerRangeValidator(0, 1));
+        VALIDATORS.put(Secure.KEYBOX_DATA, ANY_STRING_VALIDATOR);
     }
 }
-- 
2.55.0

